server/base/src/main/java/org/apache/accumulo/server/GarbageCollectionLogger.java

```diff
@@ -22,109 +22,113 @@
 import java.lang.management.ManagementFactory;
 import java.util.HashMap;
 import java.util.List;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
 
-import org.apache.accumulo.core.conf.AccumuloConfiguration;
 import org.apache.accumulo.core.conf.Property;
 import org.apache.accumulo.core.util.Halt;
+import org.apache.accumulo.core.util.threads.ThreadPools;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class GarbageCollectionLogger {
   private static final Logger log = LoggerFactory.getLogger(GarbageCollectionLogger.class);
 
+  private final ServerContext context;
+
   private final HashMap<String,Long> prevGcTime = new HashMap<>();
   private long lastMemorySize = 0;
   private long gcTimeIncreasedCount = 0;
-  private static long lastMemoryCheckTime = 0;
-  private static final Lock memCheckTimeLock = new ReentrantLock();
-
-  public void logGCInfo(AccumuloConfiguration conf) {
+  private long lastMemoryCheckTime = 0;
 
-    memCheckTimeLock.lock();
-    try {
-      final long now = System.currentTimeMillis();
+  private static final long TIME_BETWEEN_GC_CHECKS_MIN = 5;
 
-      List<GarbageCollectorMXBean> gcmBeans = ManagementFactory.getGarbageCollectorMXBeans();
-      Runtime rt = Runtime.getRuntime();
+  public GarbageCollectionLogger(final ServerContext context) {
+    this.context = context;
 
-      StringBuilder sb = new StringBuilder("gc");
+    ScheduledFuture<?> future = context.getScheduledExecutor()
+        .scheduleWithFixedDelay(this::logGCInfo, 0, TIME_BETWEEN_GC_CHECKS_MIN, TimeUnit.MINUTES);
+    ThreadPools.watchNonCriticalScheduledTask(future);
+  }
 
-      boolean sawChange = false;
+  public synchronized void logGCInfo() {
```

Comment:
I'm wondering if this method needs to be synchronized. Do we know what the lock in the previous version of the code is protecting? Is it the JMX MBeans? It looks like the only time this is called is either from the background thread, when the server is stopping, or when Halt is being invoked.

Reply:
There is a separate thread that gets started when the tserver is starting up but other threads call this method as well. I don't know why we have both the separate thread and external calls to this logger. The check for the static time was added because of an old complex issue: https://issues.apache.org/jira/browse/ACCUMULO-3084

Reply:
> I don't know why we have both the separate thread and external calls to this logger.

I'm understand why we have the background thread, but not why we are printing the JVM GC information when the server is stopped or halted.

> The check for the static time was added because of an old complex issue: https://issues.apache.org/jira/browse/ACCUMULO-3084

I see, the lock was used so that the `lastMemoryCheckTime` variable was not changed by another Thread. So, this code is trying to detect when this method is delayed due to GC. Wouldn't this cause an issue though in the case where:

1. The background thread is running, but delayed due to GC, and 
2. A call to `Halt` was issued, say by the Manager

Reply:
> this code is trying to detect when this method is delayed due to GC. Wouldn't this cause an issue

This is what I was thinking... A - It seems like this lock could cause problems. B - At best this thread _might_ have provided insight to Java GC on old versions of the JDK. C - Killing the tserver because of a guess at Java memory usage timing seems like a bad idea.

Reply:
In general - I think that this class can be removed in future versions in favor of metrics and recommending setting the appropriate  JVM options (-XX:HeapDumpOnOutOfMemoryError and/or -XX:OnOutOfMemeoryError) or whatever is appropriate for the given JVM.

However, the things that might prevent it from being removed or modified now:

- The halt in the class before an OOM is thrown may allow for a more orderly shutdown.  We should be resistant to sudden failures, but a few moments to flush / close thing might still speed recovery - but not sure of this.
- Users may be expecting these messages and/or not have the necessary metrics captured. 
- I am not sure right now if the JVM metrics alone would be sufficient, or if some variation of the values gathered / logged here should be added.

Removal / metrics changes could be another follow-on for this class.  The goal with this PR was to be an incremental improvement to resolve the synchronization issue flagged by spotbugs.  While this current iteration may not be ideal, it should at least be correct if there are concurrent log attempts whiteout significantly changing the behavior.